### DISCUSSION TITLE
Welcome to ChainLadder.jl Discussions!

### DISCUSSION THREAD

**Discussion by alecloudenback** [JuliaActuary/ChainLadder.jl#13] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by ghost** [comment]
Hi @alecloudenback I am interested in contributing to Julia could you please tell what ChainLadder.jl is about and how can I help.
I am a beginner and currently trying to learn Julia and came from https://github.com/JuliaActuary/ChainLadder.jl/issues/14 for discussion on Chainladder.jl
How can I help?
I am familiar with using Git and GitHub : )

**Reply by alecloudenback** [reply]
HI Kudeep, this would be an advanced area of contribution requiring a lot of domain expertise and Julia knowledge to help with the design and build of the package's fundamentals.

**Comment by matt-kandel** [comment]
Hey @alecloudenback I am a technical analyst at Munich Re and a beginner to the Julia language. When I was studying for the CAS Exam 5, I wrote some functions in Julia for basic reserving methods (chainladder, bornhuetter-ferguson, cape cod, etc.). I think what I have is solid, but everything is built on a Matrix{Union{Number, Missing}} type. So unlike Dataframes or Tables, it's pretty hard to slice or reference triangles by development period or accident year, or to pull data from external/tabular sources. Still, I'd like to share what I have so far and hopefully others can build off it

**Reply by alecloudenback** [reply]
Yes, please do! We can figure out how to integrate with the package.

**Reply by matt-kandel** [reply]
Cool! My project is located here: https://github.com/matt-kandel/Reserving.jl
I moved things between repos, so the tests aren't running right now, but they were passing when this project was under a different name

**Reply by alecloudenback** [reply]
i've taken a peek at your package, I don't think it would be too difficult to bring things together. I'm a life actuary without a lot of GI background besides the very basics. I think we could probably help each other - over the last few years I've gotten pretty familiar with Julia. I have a few questions/comments for your insight:

## Kinds of Triangles
It seems like data can generally be provided in what I called "incremental" and "cumulative" claims form. E.g. the former would be claims each year of `10,10,10` whereas the latter would be `10,20,30`. For that reason, I created both the Cumulative and Incremental Triangle varieties. The thinking is that depending on the algorithm, Julia's multiple dispatch will let the triangle's form tell it how to work. Or in practice do all of the algorithms just use the incremental form of the triangle and that should just be a base type?

## Data structure

The current implementations of the Triangles is a simple 3 element data structure:
  - an origin vector storing the origin times 
  - a Development vector storing the indices of the development times
  - a claims triangle the same way you have (`Union{Float64,Missing}`
  
 So I'd expect that a lot of what you've got should be pretty transportable. Currently, the origin/development indices aren't used for anything.

## Loss development algorithms

I see you've already got a handful of ways to develop the losses/reserves. I've heard it described (maybe I was reading one of the CAS study papers?) that a lot of loss development is just variations on a linear model. And I think the formulation in an `y ~ x` type of problem is something that would facilitate using MLJ.jl to drop in for reserve models the way that sci-kit does for python's chainladder package. I think here's where my lack of real experience / dedicated study had me hit a wall previously because I wasn't able to formulate the problem well enough to think about how to architect ChainLadder.jl. Maybe this explanation of what I was trying to figure out makes sense to you?

## Minimum Viable Package

What are the features that would make a package minimally viable for a practitioner? I can't really tell if ChainLadder.jl is 5% or 50% of the way to something workable.

## Collaboration

I just wanted to say again thanks for reaching out. I think Julia is just so great for actuarial work and with some domain knowledge some amazing packages can be created. However much you are interested in collaborating I'd love to learn from you.

**Reply by matt-kandel** [reply]
Thanks for the feedback. In answer to your comments:

## **Kinds of Triangles**
All of my methods so far were based on cumulative triangles, but I agree that it makes sense to split by “incremental” and “cumulative.”

## **Data Structure**
I agree with including origin/development indices in the triangle, but would it make sense to use an outer constructor to make these optional? I think some of the testing and experimentation is easier without keeping track of the indices all the time. Also, I didn’t know how to imitate the four-dimensional structure used by the Python chainladder library. At one point, I had something like this:

```
struct Triangles
    labels
    triangles::Vector{Triangle}
end
```
I wanted to have a set of checks to ensure that all triangles shared the same origin and development axes. Also, I wanted to have an arbitrary number of labels for each triangle, similar to the columns argument in Python chainladder. I never figured out a great structure though. Some kind of Vector{Dict}? Or Vector{String}? I am hopeful that we can build a lot of the same functionality from scratch. There has to be a simpler way to keep the labels synchronized with the data without depending on something like pandas.

## **Loss development algorithms**
I’m probably in the same boat as you regarding solutions based on linear models. I basically ported a spreadsheet-style method into Julia, since that’s how the material is presented to students currently. It works for now, at least, but I’m definitely open to other approaches.

## **Minimum Viable Package**
In my current role, I'm responsible for data preparation for various P&C pricing actuarial departments. I don’t know what a reserving actuary would need, since I've only seen these methods in an academic context. My own current use case is this: I want to be able to read tabular data into a triangle or set of triangles, filter/aggregate them, and display them on top of heat maps, hopefully in a format clean enough to include in a presentation or report. I’m in kind of a weird role, where I’m not really doing much analysis myself, I’m just providing the data to other people in a format they can use. However, I think if we structure these triangles right, it shouldn’t be too difficult to keep adding methods later as needed.

## **Collaboration**
It’s great to meet you! I fully agree that Julia is the right tool here. I originally got into actuarial work because I liked math, but after dealing with all the spreadsheets and insurance jargon, it’s been refreshing to program in Julia, where, as you said in a blog post somewhere, “math looks like math.” I’m not sure how much time I’ll have to contribute, between my job and studying for exams, but I’ll try to work on this when I can.

**Reply by alecloudenback** [reply]
What is the other dimension for the 4D version that you describe? A different grouping of claims like different risk cohorts?

**Reply by matt-kandel** [reply]
I'm not totally sure, but I was thinking of the Structure section in their documentation: https://chainladder-python.readthedocs.io/en/latest/triangle.html

![triangle_graphic](https://user-images.githubusercontent.com/52428107/209379978-277292bb-a4a3-4217-a98e-193e2bc4cc93.PNG)

**Reply by alecloudenback** [reply]
I added some [docstrings](https://juliaactuary.github.io/ChainLadder.jl/dev/api/) to at least bring the documentation up to date with the last iteration I had with Chainladders. 

How about this: is there a particular function from your repository that you'd like to see in ChainLadder? We can work together to port the logic over and through that I can learn how you've set up your code and you can learn ChainLadder.
